handlers/filter.go

```diff
@@ -0,0 +1,81 @@
+package handlers
+
+import (
+	"context"
+	"errors"
+	"net/url"
+	"strings"
+
+	search "github.com/ONSdigital/dp-api-clients-go/site-search"
+	"github.com/ONSdigital/dp-frontend-search-controller/data"
+	"github.com/ONSdigital/log.go/log"
+)
+
+func mapFilterTypes(ctx context.Context, query url.Values) (apiQuery url.Values, err error) {
+	apiQuery, err = url.ParseQuery(query.Encode())
+	if err != nil {
+		log.Event(ctx, "failed to parse copy of query for mapping filter types", log.Error(err), log.ERROR)
+		return nil, err
+	}
+	filters := apiQuery["filter"]
+	if len(filters) > 0 {
+		var newFilters []string
+		for _, fType := range filters {
+			found := false
+			for _, category := range data.Category {
+				for _, searchType := range category {
+					if fType == searchType.QueryType {
+						found = true
+						newFilters = append(newFilters, searchType.SubTypes)
+						break
+					}
+				}
+				if found {
+					break
+				}
+			}
+			if !found {
+				return nil, errFilterType
+			}
+		}
+		apiQuery.Del("filter")
+		apiQuery.Set("content_type", strings.Join(newFilters, ","))
+	}
+	return apiQuery, nil
+}
+
+func mapCountFilterTypes(ctx context.Context, apiQuery url.Values, searchC SearchClient) (mappedContentType []search.ContentType, err error) {
+	//Remove filter to get count of all types for the query from the client
+	apiQuery.Del("content_type")
+	countResp, err := searchC.GetSearch(ctx, apiQuery)
+	if err != nil {
+		log.Event(ctx, "getting search query count from client failed", log.Error(err), log.ERROR)
+		return nil, err
+	}
+	countFilter := make(map[string]int)
+	for _, contentType := range countResp.ContentTypes {
+		foundFilter := false
+		for _, category := range data.Category {
+			for _, searchType := range category {
+				mapfilters := strings.Split(searchType.SubTypes, ",")
```

Comment:
**Question:** As the category is a struct we've created for this is there a reason that the subtypes is a `string` instead of a `[]string`? The name implies plural instead of a single comma separated value.

Generally speaking a struct should provide useful methods for the data it holds i.e. providing it in the format required by its users.  If there is a need for this to be a CSV string then consider adding a method to `Category` to get `SubTypes` as a `[]string`. This will simplify the calling code as it can get the data in the format it needs from the struct rather than having to transforms it all the time.

Reply:
The subtypes were initially `string` because it needs to be passed to the logic (dp-search-query - `GetSearch` method) in a `string` form.